- This notebook aims to find a way to pre-process transcripts within the given dataframe

# Step-1 Create a randomly sampled dataframe

In [2]:
import pandas as pd

## Step-1.1 Data Ingestion

In [3]:
# Read the eng_subtitles dataframe
df=pd.read_csv('../data/eng_subtitles.csv')
print(df.info())
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 82498 entries, 0 to 82497
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   num              82498 non-null  int64 
 1   name             82498 non-null  object
 2   decoded_content  82498 non-null  object
dtypes: int64(1), object(2)
memory usage: 1.9+ MB
None


num                                               name  \
0  9180533                         the.message.(1976).eng.1cd   
1  9180583  here.comes.the.grump.s01.e09.joltin.jack.in.bo...   
2  9180592    yumis.cells.s02.e13.episode.2.13.(2022).eng.1cd   
3  9180594    yumis.cells.s02.e14.episode.2.14.(2022).eng.1cd   
4  9180600                              broker.(2022).eng.1cd   

                                     decoded_content  
0  1\r\n00:00:06,000 --> 00:00:12,074\r\nWatch an...  
1  1\r\n00:00:29,359 --> 00:00:32,048\r\nAh! Ther...  
2  1\r\n00:00:53,200 --> 00:00:56,030\r\n<i>Yumi'...  
3  1\r\n00:00:06,000 --> 00:00:12,074\r\nWatch an...  
4  1\r\n00:00:06,000 --> 00:00:12,074\r\nWatch...

In [4]:
#df=df.drop('num',axis=1)
#df['index']=df.index
#df=df.set_index('index')
df

name  \
index                                                      
0                             the.message.(1976).eng.1cd   
1      here.comes.the.grump.s01.e09.joltin.jack.in.bo...   
2        yumis.cells.s02.e13.episode.2.13.(2022).eng.1cd   
3        yumis.cells.s02.e14.episode.2.14.(2022).eng.1cd   
4                                  broker.(2022).eng.1cd   
...                                                  ...   
82493                   the.prophets.game.(2000).eng.1cd   
82494                         west.beirut.(1998).eng.1cd   
82495         frankenstein.the.true.story.(1973).eng.1cd   
82496         frankenstein.the.true.story.(1973).eng.1cd   
82497              zombie.island.massacre.(1984).eng.1cd   

                                         decoded_content  
index                                                     
0      1\r\n00:00:06,000 --> 00:00:12,074\r\nWatch an...  
1      1\r\n00:00:29,359 --> 00:00:32,048\r\nAh! Ther...  
2      1\r\n00:00:53,200 --> 00:00:56,030\r\n<i>Yumi'...  
3      1\r\n00:00:06,000 --> 00:00:12,074\r\nWatch an...  
4      1\r\n00:00:06,000 --> 00:00:12,074\r\nWatch...  
...                                                  ...  
82493  1\r\n00:01:16,284 --> 00:01:19,537\r\nGod,\...  
82494  1\r\n00:00:06,000 --> 00:00:12,074\r\napi.Open...  
82495  1\r\n00:00:01,001 --> 00:00:04,630\r\n(Dramati...  
82496  1\r\n00:00:06,000 --> 00:00:12,074\r\nAdvertis...  
82497  1\r\n00:00:01,919 --> 00:00:03,253\r\n(Sharp w...  

[82498 rows x 2 columns]

## Step-1.2 Deriving a random sample from whole dataframe

In [6]:
import numpy as np
import random
random.seed(42)

random_ind=np.random.randint(0,len(df),1000,)
random_ind[:5]

array([50821, 30235, 34474, 37802, 74591])

In [7]:
df=df.iloc[random_ind,:]
df

name  \
index                                                      
50821     law.order.s22.e11.second.chance.(2023).eng.1cd   
30235   bergerac.s07.e07.old.acquaintance.(1989).eng.1cd   
34474  seiken.densetsu.legend.of.mana.the.teardrop.cr...   
37802                           hard.quiz.(2016).eng.1cd   
74591  being.human.s02.e04.(i.loathe.you).for.sentime...   
...                                                  ...   
6175                         4th.republic.(2019).eng.1cd   
12986  bucchigire.s01.e07.prevent.the.dissolution.of....   
11464  naomi.s01.e01.dont.believe.everything.you.thin...   
56794  family.guy.s04.e16.the.courtship.of.stewies.fa...   
45694  the.king.of.queens.s07.e12.gym.neighbors.(2005...   

                                         decoded_content  
index                                                     
50821  1\r\n00:00:10,545 --> 00:00:12,272\r\n<i>- ...  
30235  1\r\n00:00:06,000 --> 00:00:12,074\r\napi.O...  
34474  [Script Info]\r\nTitle: English (US)\r\nOrigin...  
37802  1\r\n00:00:02,140 --> 00:00:03,460\r\nANNOU...  
74591  1\r\n00:00:00,868 --> 00:00:02,803\r\n- Pre...  
...                                                  ...  
6175   1\r\n00:00:06,000 --> 00:00:12,074\r\nWatch...  
12986  [Script Info]\r\nTitle: English (US)\r\nOrigin...  
11464  1\r\n00:00:02,086 --> 00:00:03,338\r\n[Tkay...  
56794  1\r\n00:00:03,170 --> 00:00:06,004\r\n♪ I...  
45694  1\r\n00:00:07,590 --> 00:00:09,957\r\n<i>- all...  

[1000 rows x 2 columns]

# Step-2 Pre-processing content

## Step-2.1 Observing possible formats of given transcripts

In [8]:
index=df.index

In [9]:
print(df['decoded_content'][index[0]])

1
00:00:10,545 --> 00:00:12,272
<i>- In the criminal
justice system,</i>

2
00:00:12,297 --> 00:00:13,632
<i>the people are represented</i>

3
00:00:13,657 --> 00:00:15,993
<i>by two separate, yet
equally important groups:</i>

4
00:00:16,018 --> 00:00:18,107
<i>the police, who
investigate crime,</i>

5
00:00:18,132 --> 00:00:20,968
<i>and the district attorneys,
who prosecute the offenders.</i>

6
00:00:20,993 --> 00:00:22,731
<i>These are their stories.</i>

7
00:00:24,000 --> 00:00:30,074
api.OpenSubtitles.org is deprecated, please
implement REST API from OpenSubtitles.com

8
00:00:43,626 --> 00:00:47,130
- Yo.

9
00:00:47,172 --> 00:00:49,215
Yo.

10
00:01:04,022 --> 00:01:05,148
Hey!

11
00:01:06,983 --> 00:01:09,736
Yo, what you running for?

12
00:01:09,778 --> 00:01:11,321
Hey!

13
00:01:13,740 --> 00:01:15,825
- Early bird runner
found him laying here.

14
00:01:15,867 --> 00:01:17,160
- Approximate time of death?

15
00:01:17,202 --> 00:01:19,245
- MLIs put it somewhere
- 

In [16]:
print(df['decoded_content'][index[1]])

1
00:00:09,727 --> 00:00:13,629
WHEN I BEGAN MY FIELD RESEARCH
HERE, THESE FROGS WERE THRIVING.

2
00:00:13,697 --> 00:00:17,766
SINCE THEN, THE ADULT POPULATION
HAS DROPPED TO FEWER THAN 200.

3
00:00:17,835 --> 00:00:22,521
IF YOU DON'T TAKE
PROTECTIVE MEASURES,
IN A FEW YEARS...

4
00:00:22,590 --> 00:00:24,990
<i>RHINUS VENISEFULA</i> WILL BE
EXTINCT HERE.

5
00:00:27,244 --> 00:00:32,531
FROG POPULATIONS ARE
DECLINING ALL OVER THE GLOBE,
DR. FARADAY. NO ONE KNOWS WHY.

6
00:00:32,600 --> 00:00:35,884
WE CAN'T POSSIBLY PLACE THEM ALL
ON THE ENDANGERED SPECIES LIST.

7
00:00:35,953 --> 00:00:39,621
WE'D FIND A WAY IF IT WERE
CUTE, FURRY LITTLE MAMMALS
WE WERE TALKING ABOUT.

8
00:00:39,690 --> 00:00:44,426
REGARDLESS, YOUR STUDY
REMAINS INCONCLUSIVE...
SPECULATIVE AT BEST.

9
00:00:44,495 --> 00:00:46,445
YOU'VE PROVIDED
NO CONCRETE EVIDENCE...

10
00:00:46,513 --> 00:00:50,114
THE FROG DEPOPULATION
IS THE EXCLUSIVE RESULT
OF HUMAN ENCROACHMENT.

11
00:00:50,183 --> 00:00:54,869


In [10]:
print(df['decoded_content'][index[2]])

[Script Info]
Title: English (US)
Original Script: WarnerJapan  [http://www.crunchyroll.com/user/WarnerJapan]
Original Translation: 
Original Editing: 
Original Timing: 
Synch Point: 
Script Updated By: 
Update Details: 
ScriptType: v4.00+
Collisions: Normal
PlayResX: 640
PlayResY: 360
Timer: 0.0000
WrapStyle: 0

[V4+ Styles]
Format: Name,Fontname,Fontsize,PrimaryColour,SecondaryColour,OutlineColour,BackColour,Bold,Italic,Underline,Strikeout,ScaleX,ScaleY,Spacing,Angle,BorderStyle,Outline,Shadow,Alignment,MarginL,MarginR,MarginV,Encoding
Style: Default,Arial,20,&H00FFFFFF,&H000000FF,&H00000000,&H00000000,0,0,0,0,100,100,0,0,1,2,2,2,0010,0010,0010,1
Style: Main,Trebuchet MS,24,&H00FFFFFF,&H000000FF,&H00000000,&H00000000,0,0,0,0,100,100,0,0,1,2,1,2,0010,0010,0018,1
Style: Italics,Trebuchet MS,24,&H00FFFFFF,&H000000FF,&H00000000,&H00000000,0,-1,0,0,100,100,0,0,1,2,1,2,0010,0010,0018,1
Style: Flashback,Trebuchet MS,24,&H00FFFFFF,&H000000FF,&H00500000,&H00500000,0,0,0,0,100,100,0,0,1,2,1,2,

### Observation
- Looks like there are two types of transcripts
    - Anime
    - Non- Anime
- Both have different formats. 
- Need to create a mechanism to to identify which script it is and pre-preocessing them

## Step-2.2 Creating strategies to pre-process the movies and tv shows transcripts

In [51]:
# Let's try out the pattern on a sample timestamp
import re
timestamp_pattern=r'\d+\r?\n\d{2}:\d{2}:\d{2},\d{3} --> \d{2}:\d{2}:\d{2},\d{3}'
text=df['decoded_content'][index[0]]#movie
text1=re.sub(timestamp_pattern,'',text)
print('After remvoing Timestamps:',text1)



After remvoing Timestamps: 
<i>- In the criminal
justice system,</i>


<i>the people are represented</i>


<i>by two separate, yet
equally important groups:</i>


<i>the police, who
investigate crime,</i>


<i>and the district attorneys,
who prosecute the offenders.</i>


<i>These are their stories.</i>


api.OpenSubtitles.org is deprecated, please
implement REST API from OpenSubtitles.com


- Yo.


Yo.


Hey!


Yo, what you running for?


Hey!


- Early bird runner
found him laying here.


- Approximate time of death?


- MLIs put it somewhere
- between 2:00 and 4:00 AM.


- Witnesses?
- None so far.


- Yeah, nobody
sees nothing, huh?


- Eh.


Blunt force trauma?


- That'd be my guess.


- ID?


- Derek Stanton.


- Still got his cash.


Any indication that the
body was moved here?


- The way the blood has
pooled and the spatter


on the pavement, I think
it all happened right here.


- My boy had quite
the rap sheet.


Just got out of Fishkill
nine months ago.


- What was he 

- Note: The below step was performed, because some movie and tv show transcripts have header in them. But the current example doesn't have it

In [52]:
#removing header
header_pattern1=r'(\r\nWatch any video online with Open-SUBTITLES\r\nFree Browser extension: osdb.link/ext\r\n\r\n\r\n)|(\r\nAdvertise your product or brand here\r\ncontact www.OpenSubtitles.org today)'#header
text2=re.sub(header_pattern1,'',text1)
print('\nAfter removing header_text:',text2)



After removing header_text: 
<i>- In the criminal
justice system,</i>


<i>the people are represented</i>


<i>by two separate, yet
equally important groups:</i>


<i>the police, who
investigate crime,</i>


<i>and the district attorneys,
who prosecute the offenders.</i>


<i>These are their stories.</i>


api.OpenSubtitles.org is deprecated, please
implement REST API from OpenSubtitles.com


- Yo.


Yo.


Hey!


Yo, what you running for?


Hey!


- Early bird runner
found him laying here.


- Approximate time of death?


- MLIs put it somewhere
- between 2:00 and 4:00 AM.


- Witnesses?
- None so far.


- Yeah, nobody
sees nothing, huh?


- Eh.


Blunt force trauma?


- That'd be my guess.


- ID?


- Derek Stanton.


- Still got his cash.


Any indication that the
body was moved here?


- The way the blood has
pooled and the spatter


on the pavement, I think
it all happened right here.


- My boy had quite
the rap sheet.


Just got out of Fishkill
nine months ago.


- What was h

In [53]:
# remvoing footer_text
footer_pattern=r'(Please rate this subtitle at www.osdb.link/[0-9a-z]+\r\nHelp other users to choose the best subtitles)' #footer
text3=re.sub(footer_pattern,'',text2)
print('\nAfter removing footer_text:',text3)


After removing footer_text: 
<i>- In the criminal
justice system,</i>


<i>the people are represented</i>


<i>by two separate, yet
equally important groups:</i>


<i>the police, who
investigate crime,</i>


<i>and the district attorneys,
who prosecute the offenders.</i>


<i>These are their stories.</i>


api.OpenSubtitles.org is deprecated, please
implement REST API from OpenSubtitles.com


- Yo.


Yo.


Hey!


Yo, what you running for?


Hey!


- Early bird runner
found him laying here.


- Approximate time of death?


- MLIs put it somewhere
- between 2:00 and 4:00 AM.


- Witnesses?
- None so far.


- Yeah, nobody
sees nothing, huh?


- Eh.


Blunt force trauma?


- That'd be my guess.


- ID?


- Derek Stanton.


- Still got his cash.


Any indication that the
body was moved here?


- The way the blood has
pooled and the spatter


on the pavement, I think
it all happened right here.


- My boy had quite
the rap sheet.


Just got out of Fishkill
nine months ago.


- What was h

- Note: Some transcripts had links within them. Current example don't have links within them

In [54]:
#removing links
link_pattern=r'(https?:\/\/)?(www\.)?(\w+)(\.\w+)'
text4=re.sub(link_pattern,'',text3)
print('After removing links:',text4)

After removing links: 
<i>- In the criminal
justice system,</i>


<i>the people are represented</i>


<i>by two separate, yet
equally important groups:</i>


<i>the police, who
investigate crime,</i>


<i>and the district attorneys,
who prosecute the offenders.</i>


<i>These are their stories.</i>


.org is deprecated, please
implement REST API from 


- Yo.


Yo.


Hey!


Yo, what you running for?


Hey!


- Early bird runner
found him laying here.


- Approximate time of death?


- MLIs put it somewhere
- between 2:00 and 4:00 AM.


- Witnesses?
- None so far.


- Yeah, nobody
sees nothing, huh?


- Eh.


Blunt force trauma?


- That'd be my guess.


- ID?


- Derek Stanton.


- Still got his cash.


Any indication that the
body was moved here?


- The way the blood has
pooled and the spatter


on the pavement, I think
it all happened right here.


- My boy had quite
the rap sheet.


Just got out of Fishkill
nine months ago.


- What was he in for?


- Selling marijuana.


Served

- Note: Some transcripts had html tags within them. Current example don't have it

In [55]:
# remove html tags
html_pattern=r'<[/]?\w+>'
text5=re.sub(html_pattern,'',text4)
print('After removing html tags:',text5)

After removing html tags: 
- In the criminal
justice system,


the people are represented


by two separate, yet
equally important groups:


the police, who
investigate crime,


and the district attorneys,
who prosecute the offenders.


These are their stories.


.org is deprecated, please
implement REST API from 


- Yo.


Yo.


Hey!


Yo, what you running for?


Hey!


- Early bird runner
found him laying here.


- Approximate time of death?


- MLIs put it somewhere
- between 2:00 and 4:00 AM.


- Witnesses?
- None so far.


- Yeah, nobody
sees nothing, huh?


- Eh.


Blunt force trauma?


- That'd be my guess.


- ID?


- Derek Stanton.


- Still got his cash.


Any indication that the
body was moved here?


- The way the blood has
pooled and the spatter


on the pavement, I think
it all happened right here.


- My boy had quite
the rap sheet.


Just got out of Fishkill
nine months ago.


- What was he in for?


- Selling marijuana.


Served four years.


- Four years for sellin

In [56]:
# removing extra line breaks
linebreak_pattern=r'(\r\n)+'
text6=re.sub(linebreak_pattern,'\r\n',text5)
print('After removing extra line breaks:',text6)

After removing extra line breaks: 
- In the criminal
justice system,
the people are represented
by two separate, yet
equally important groups:
the police, who
investigate crime,
and the district attorneys,
who prosecute the offenders.
These are their stories.
.org is deprecated, please
implement REST API from 
- Yo.
Yo.
Hey!
Yo, what you running for?
Hey!
- Early bird runner
found him laying here.
- Approximate time of death?
- MLIs put it somewhere
- between 2:00 and 4:00 AM.
- Witnesses?
- None so far.
- Yeah, nobody
sees nothing, huh?
- Eh.
Blunt force trauma?
- That'd be my guess.
- ID?
- Derek Stanton.
- Still got his cash.
Any indication that the
body was moved here?
- The way the blood has
pooled and the spatter
on the pavement, I think
it all happened right here.
- My boy had quite
the rap sheet.
Just got out of Fishkill
nine months ago.
- What was he in for?
- Selling marijuana.
Served four years.
- Four years for selling weed?
- Crazy.
- Times have changed.
Surveillance fo

### Observation
- Every transcript in the database, has 
    - timestamps
    - header and footer
    - html tags
    - links
    - multiple line breaks(\r\n)
- As part of preprocessing, we need to remove them from the text


In [57]:
# Let's try the pattern on a sample row
text_df=df['decoded_content'][index[0]]
text1=re.sub(timestamp_pattern,'',text_df)
text2=re.sub(header_pattern1,'',text1)
text3=re.sub(footer_pattern,'',text2)
text4=re.sub(link_pattern,'',text3)
text5=re.sub(html_pattern,'',text4)#removing html tags
text6=re.sub(linebreak_pattern,'\r\n',text5)
#print('After remvoing Timestamps:',text1)
# print('After removing header_text:',text2)
print('cleaned_content:',text6)

cleaned_content: 
- In the criminal
justice system,
the people are represented
by two separate, yet
equally important groups:
the police, who
investigate crime,
and the district attorneys,
who prosecute the offenders.
These are their stories.
.org is deprecated, please
implement REST API from 
- Yo.
Yo.
Hey!
Yo, what you running for?
Hey!
- Early bird runner
found him laying here.
- Approximate time of death?
- MLIs put it somewhere
- between 2:00 and 4:00 AM.
- Witnesses?
- None so far.
- Yeah, nobody
sees nothing, huh?
- Eh.
Blunt force trauma?
- That'd be my guess.
- ID?
- Derek Stanton.
- Still got his cash.
Any indication that the
body was moved here?
- The way the blood has
pooled and the spatter
on the pavement, I think
it all happened right here.
- My boy had quite
the rap sheet.
Just got out of Fishkill
nine months ago.
- What was he in for?
- Selling marijuana.
Served four years.
- Four years for selling weed?
- Crazy.
- Times have changed.
Surveillance footage from
the pa

## Step-2.3 Creating a function to pre-process movie and tv show transcripts

In [58]:
import nltk
#nltk.download('punkt')
#nltk.download('wordnet')
from nltk.tokenize import word_tokenize
from nltk.stem import SnowballStemmer
import re
def text_preprocessing_non_anime(corpus,flag):
    
    #remove timestamps
    timestamp_pattern=r'\d+\r?\n\d{2}:\d{2}:\d{2},\d{3} --> \d{2}:\d{2}:\d{2},\d{3}'
    corpus=re.sub(timestamp_pattern,'',corpus)
    
    #remov header and footer
    header_pattern1=r'(\r\nWatch any video online with Open-SUBTITLES\r\nFree Browser extension: osdb.link/ext\r\n\r\n\r\n)|(\r\nAdvertise your product or brand here\r\ncontact www.OpenSubtitles.org today)'#header
    corpus=re.sub(header_pattern1,'',corpus)
    footer_pattern=r'Please rate this subtitle at www.osdb.link/[0-9a-z]+\r\nHelp other users to choose the best subtitles' #footer
    corpus=re.sub(footer_pattern,'',corpus)

    #removing links
    link_pattern=r'(https?:\/\/)?(www\.)?(\w+)(\.\w+)'
    text4=re.sub(link_pattern,'',text3)
    
    #remove html tags
    html_pattern=r'<[/]?\w+>'
    corpus=re.sub(html_pattern,'',corpus)
    
    #remove extra line breaks
    linebreak_pattern=r'(\r\n)+'
    corpus=re.sub(linebreak_pattern,'\r\n',corpus)
    
    
    # remove special characters
    corpus=re.sub('[^a-zA-Z]',' ',corpus)
    
    #convert to lower case
    corpus=corpus.lower()

    #tokenize
    words=word_tokenize(corpus)
    if flag=="stemming":
    #stemming
        stemmer=SnowballStemmer(language='english')
        return ' '.join(stemmer.stem(word) for word in words if word not in set(nltk.corpus.stopwords.words('english')))
    else:
    #lemmatization
        lemmatizer=WordNetLemmatizer()
        return ' '.join(lemmatizer.lemmatize(word) for word in words if word not in set(nltk.corpus.stopwords.words('english')))


#flag is either "stemming" or "lemmatization"


Note: It is evident by observing below result that we cannot remove 100% noise from text. Because, there are still some random comments within transcripts. Our aim is to minimize noise. And we are successful to limit it to an extent

In [59]:
#sending a single transcript
print(text_preprocessing_non_anime(df['decoded_content'][index[1]],"stemming"))

api opensubtitl org deprec pleas implement rest api opensubtitl com help sir er yes expect christin nash ah yes like follow peter welcom back london look marvel alway thank oslo great order bottl right oh fine waiter would like order madam order littl met far know littl smokescreen hope mind call see billi car last night includ two porsch bmw leav till gone three clock morn told must drunk terribl thing drink ask thank much billi spent last night virtuous isol say might one two mate jar er car mate bit hard say like yeah well strain shall see left place tidi said phone busi proposit right peter lucrat proposit work selco intern research show sizeabl gap soft drink market way see sticki self indulg hard healthi look somebodi alreadi imag somebodi sell product simpli associ six year sinc broke record medal well still univers respect figur peter world athlet public larg well known straight talk integr cours accept offer want start fair soon mind mean want start shoot pilot commerci almost

## Step -2.4 Pre-processing Anime Transcripts

In [60]:
df['decoded_content'][index[2]]

'[Script Info]\r\nTitle: English (US)\r\nOriginal Script: WarnerJapan  [http://www.crunchyroll.com/user/WarnerJapan]\r\nOriginal Translation: \r\nOriginal Editing: \r\nOriginal Timing: \r\nSynch Point: \r\nScript Updated By: \r\nUpdate Details: \r\nScriptType: v4.00+\r\nCollisions: Normal\r\nPlayResX: 640\r\nPlayResY: 360\r\nTimer: 0.0000\r\nWrapStyle: 0\r\n\r\n[V4+ Styles]\r\nFormat: Name,Fontname,Fontsize,PrimaryColour,SecondaryColour,OutlineColour,BackColour,Bold,Italic,Underline,Strikeout,ScaleX,ScaleY,Spacing,Angle,BorderStyle,Outline,Shadow,Alignment,MarginL,MarginR,MarginV,Encoding\r\nStyle: Default,Arial,20,&H00FFFFFF,&H000000FF,&H00000000,&H00000000,0,0,0,0,100,100,0,0,1,2,2,2,0010,0010,0010,1\r\nStyle: Main,Trebuchet MS,24,&H00FFFFFF,&H000000FF,&H00000000,&H00000000,0,0,0,0,100,100,0,0,1,2,1,2,0010,0010,0018,1\r\nStyle: Italics,Trebuchet MS,24,&H00FFFFFF,&H000000FF,&H00000000,&H00000000,0,-1,0,0,100,100,0,0,1,2,1,2,0010,0010,0018,1\r\nStyle: Flashback,Trebuchet MS,24,&H00FFFF

In [61]:
print(df['decoded_content'][index[2]])

[Script Info]
Title: English (US)
Original Script: WarnerJapan  [http://www.crunchyroll.com/user/WarnerJapan]
Original Translation: 
Original Editing: 
Original Timing: 
Synch Point: 
Script Updated By: 
Update Details: 
ScriptType: v4.00+
Collisions: Normal
PlayResX: 640
PlayResY: 360
Timer: 0.0000
WrapStyle: 0

[V4+ Styles]
Format: Name,Fontname,Fontsize,PrimaryColour,SecondaryColour,OutlineColour,BackColour,Bold,Italic,Underline,Strikeout,ScaleX,ScaleY,Spacing,Angle,BorderStyle,Outline,Shadow,Alignment,MarginL,MarginR,MarginV,Encoding
Style: Default,Arial,20,&H00FFFFFF,&H000000FF,&H00000000,&H00000000,0,0,0,0,100,100,0,0,1,2,2,2,0010,0010,0010,1
Style: Main,Trebuchet MS,24,&H00FFFFFF,&H000000FF,&H00000000,&H00000000,0,0,0,0,100,100,0,0,1,2,1,2,0010,0010,0018,1
Style: Italics,Trebuchet MS,24,&H00FFFFFF,&H000000FF,&H00000000,&H00000000,0,-1,0,0,100,100,0,0,1,2,1,2,0010,0010,0018,1
Style: Flashback,Trebuchet MS,24,&H00FFFFFF,&H000000FF,&H00500000,&H00500000,0,0,0,0,100,100,0,0,1,2,1,2,

### Observation
- There is only header in anime transcripts

In [62]:

# Assuming text contains your input text
text=df['decoded_content'][index[2]]

# Define the regular expression pattern to match the header content
header_pattern = r'\[Script Info\].*?\[Events\]\r\n.*?\r\n'

# Remove the header content using the regular expression
cleaned_text = re.sub(header_pattern, '', text, flags=re.DOTALL)

# Print the cleaned text
print(cleaned_text)

Dialogue: 0,0:00:07.25,0:00:09.24,Main,Pearl,0000,0000,0000,,Look, Elazul!
Dialogue: 0,0:00:09.86,0:00:11.57,Main,Pearl,0000,0000,0000,,The sea!
Dialogue: 0,0:00:11.81,0:00:13.88,Main,Elazul,0000,0000,0000,,Yeah. It's pretty.
Dialogue: 0,0:00:13.88,0:00:16.03,Main,Seraphina,0000,0000,0000,,Is this your first time seeing the sea?
Dialogue: 0,0:00:16.33,0:00:18.94,Main,Pearl,0000,0000,0000,,No, I've seen it from a distance before.
Dialogue: 0,0:00:18.94,0:00:21.62,Main,Pearl,0000,0000,0000,,But not from up-close.
Dialogue: 0,0:00:23.15,0:00:25.12,Main,Seraphina,0000,0000,0000,,Shall we take a break, then?
Dialogue: 0,0:00:27.32,0:00:28.55,Main,Esmeralda,0000,0000,0000,,Big Sis...
Dialogue: 0,0:00:28.55,0:00:30.55,Main,Esmeralda,0000,0000,0000,,I sense Big Sis' glimmer.
Dialogue: 0,0:00:30.79,0:00:31.68,Main,Elazul,0000,0000,0000,,Yeah.
Dialogue: 0,0:00:31.68,0:00:33.91,Main,Elazul,0000,0000,0000,,I sense it faintly, too.
Dialogue: 0,0:00:33.91,0:00:35.13,Main,Pearl,0000,0000,0000,,I do, 

In [63]:
# remove dialogue suffix before each sentence

# Define the regular expression pattern to match the specified format
dialogue_pattern =r"Dialogue: (Marked=)?\d+,\d{1,2}:\d{2}:\d{2}\.\d{2},\d{1,2}:\d{2}:\d{2}\.\d{2},\*?(\w+)?,(\w+)?,\d{1,4},\d{1,4},\d{1,4},,"

cleaned_text=re.sub(dialogue_pattern,'',cleaned_text)
print(cleaned_text)

Look, Elazul!
The sea!
Yeah. It's pretty.
Is this your first time seeing the sea?
No, I've seen it from a distance before.
But not from up-close.
Shall we take a break, then?
Big Sis...
I sense Big Sis' glimmer.
Yeah.
I sense it faintly, too.
I do, too.
Shiloh, let's go down to the beach.
Okay, let's do that.
{\fad(500,550)\bord1.5\fsp2}Diamond
It's so pretty.
What is it?
Pearl?
Crabs?
Look at all of them.
There's more of them than \Nlast time I was here.
You've been here, Shiloh?
Yeah, before I went to Polpota.
Big Sis...
Well, Esmeralda?
Yeah, I think she's around here somewhere.
We're definitely closer.
Do you think she's in the sand?
We should probably narrow down our search.
Let's walk a bit.
Okay.
Well?
This way?
I think we're further now.
Dialogue: 0,0:03:06.45,0:03:07.61,Main - Top,Shiloh,0000,0000,0000,,Okay.
Dialogue: 0,0:03:07.61,0:03:09.34,Main - Top,Shiloh,0000,0000,0000,,Let's go the other way.
Is that Shiloh-kun's voice?
Shiloh-kun!
Valerie.
You know her?
Yeah.
I'm gonna

In [64]:
# remove subtitle word
cleaned_text=re.sub('Subtitle:','',cleaned_text)
cleaned_text

'Look, Elazul!\r\nThe sea!\r\nYeah. It\'s pretty.\r\nIs this your first time seeing the sea?\r\nNo, I\'ve seen it from a distance before.\r\nBut not from up-close.\r\nShall we take a break, then?\r\nBig Sis...\r\nI sense Big Sis\' glimmer.\r\nYeah.\r\nI sense it faintly, too.\r\nI do, too.\r\nShiloh, let\'s go down to the beach.\r\nOkay, let\'s do that.\r\n{\\fad(500,550)\\bord1.5\\fsp2}Diamond\r\nIt\'s so pretty.\r\nWhat is it?\r\nPearl?\r\nCrabs?\r\nLook at all of them.\r\nThere\'s more of them than \\Nlast time I was here.\r\nYou\'ve been here, Shiloh?\r\nYeah, before I went to Polpota.\r\nBig Sis...\r\nWell, Esmeralda?\r\nYeah, I think she\'s around here somewhere.\r\nWe\'re definitely closer.\r\nDo you think she\'s in the sand?\r\nWe should probably narrow down our search.\r\nLet\'s walk a bit.\r\nOkay.\r\nWell?\r\nThis way?\r\nI think we\'re further now.\r\nDialogue: 0,0:03:06.45,0:03:07.61,Main - Top,Shiloh,0000,0000,0000,,Okay.\r\nDialogue: 0,0:03:07.61,0:03:09.34,Main - Top,Sh

## Step-2.5 Creating a anime transcript pre-processing function

In [65]:
import nltk
#nltk.download('punkt')
#nltk.download('wordnet')
from nltk.tokenize import word_tokenize
from nltk.stem import SnowballStemmer
import re
def text_preprocessing_anime(corpus,flag):
    
    #removing header
    header_pattern = r'\[Script Info\].*?\[Events\]\r\n.*?\r\n'
    corpus=re.sub(header_pattern,'',corpus,flags=re.DOTALL)
    
    #removing dialogue
    dialogue_pattern = r"Dialogue: (Marked=)?\d+,\d{1,2}:\d{2}:\d{2}\.\d{2},\d{1,2}:\d{2}:\d{2}\.\d{2},\*?(\w+)?,(\w+)?,\d{1,4},\d{1,4},\d{1,4},,"
    corpus=re.sub(dialogue_pattern,'',corpus)
    
    #removing subtitle
    subtitle_pattern = r'Subtitle: '
    corpus=re.sub(subtitle_pattern,'',corpus)
    
    #removing links
    pattern4=r'(https?:\/\/)?(www\.)?(\w+)(\.\w+)'
    corpus=re.sub(pattern4,'',corpus)
    
    #remove html tags
    pattern5=r'<[/]?\w+>'
    corpus=re.sub(pattern5,'',corpus)
    
    #remove extra line breaks
    pattern6=r'(\r\n)+'
    corpus=re.sub(pattern6,'\r\n',corpus)
    
    
    # remove special characters
    corpus=re.sub('[^a-zA-Z]',' ',corpus)
    
    #convert to lower case
    corpus=corpus.lower()

    #tokenize
    words=word_tokenize(corpus)
    if flag=="stemming":
    #stemming
        stemmer=SnowballStemmer(language='english')
        return ' '.join(stemmer.stem(word) for word in words if word not in set(nltk.corpus.stopwords.words('english')))
    else:
    #lemmatization
        lemmatizer=WordNetLemmatizer()
        return ' '.join(lemmatizer.lemmatize(word) for word in words if word not in set(nltk.corpus.stopwords.words('english')))


#flag is either "stemming" or "lemmatization"


In [66]:
text_preprocessing_anime(df['decoded_content'][index[2]],"stemming")

'look elazul sea yeah pretti first time see sea seen distanc close shall take break big sis sens big sis glimmer yeah sens faint shiloh let go beach okay let fad fsp diamond pretti pearl crab look nlast time shiloh yeah went polpota big sis well esmeralda yeah think around somewher definit closer think sand probabl narrow search let walk bit okay well way think dialogu main top shiloh okay dialogu main top shiloh let go way shiloh kun voic shiloh kun valeri know yeah gon na go say hi long time valeri come swim friend travel search someth egg huh yes congratul could egg fruit love david dialogu main esmeralda seraphina david yes david belov valeri betray captain return ship belov david sorri freedom na lowli pirat like right go go defeat everi last monster nso never attack sea david make even difficult stop valeri come back mayb ten year twenti year hell might hundr year come back captain nof hundr penguin pirat long belov david warm egg may happi hope david get home yes noth fell love 

## Step-2.6 Creating a Pre-processing function which can handle both

In [67]:
def text_preprocessing(corpus,flag):
    
    timestamp_pattern=r'\d+\r?\n\d{2}:\d{2}:\d{2},\d{3} --> \d{2}:\d{2}:\d{2},\d{3}'
    dialogue_pattern = r"Dialogue: (Marked=)?\d+,\d{1,2}:\d{2}:\d{2}\.\d{2},\d{1,2}:\d{2}:\d{2}\.\d{2},\*?(\w+)?,(\w+)?,\d{1,4},\d{1,4},\d{1,4},,"
    
    if re.search(timestamp_pattern,corpus): #movie or tv show
        #remove timestamps
        corpus=re.sub(timestamp_pattern,'',corpus)
        
        #remove header and footer
        header_pattern1=r'(\r\nWatch any video online with Open-SUBTITLES\r\nFree Browser extension: osdb.link/ext\r\n\r\n\r\n)|(\r\nAdvertise your product or brand here\r\ncontact www.OpenSubtitles.org today)'#header
        corpus=re.sub(header_pattern1,'',corpus)
        footer_pattern=r'Please rate this subtitle at www.osdb.link/[0-9a-z]+\r\nHelp other users to choose the best subtitles' #footer
        corpus=re.sub(footer_pattern,'',corpus)
    
    else: #anime
         #removing header
        header_pattern = r'\[Script Info\].*?\[Events\]\r\n.*?\r\n'
        corpus=re.sub(header_pattern,'',corpus,flags=re.DOTALL)
        
        #removing dialogue
        corpus=re.sub(dialogue_pattern,'',corpus)
        
        #removing subtitle
        subtitle_pattern = r'Subtitle: '
        corpus=re.sub(subtitle_pattern,'',corpus)
    

    #removing links
    link_pattern=r'(https?:\/\/)?(www\.)?(\w+)(\.\w+)'
    text4=re.sub(link_pattern,'',text3)
    
    #remove html tags
    html_pattern=r'<[/]?\w+>'
    corpus=re.sub(html_pattern,'',corpus)
    
    #remove extra line breaks
    linebreak_pattern=r'(\r\n)+'
    corpus=re.sub(linebreak_pattern,'\r\n',corpus)
    
    
    # remove special characters
    corpus=re.sub('[^a-zA-Z]',' ',corpus)
    
    #convert to lower case
    corpus=corpus.lower()

    #tokenize
    words=word_tokenize(corpus)
    if flag=="stemming":
    #stemming
        stemmer=SnowballStemmer(language='english')
        return ' '.join(stemmer.stem(word) for word in words if word not in set(nltk.corpus.stopwords.words('english')))
    else:
    #lemmatization
        lemmatizer=WordNetLemmatizer()
        return ' '.join(lemmatizer.lemmatize(word) for word in words if word not in set(nltk.corpus.stopwords.words('english')))


    
    

In [68]:
text_preprocessing(df['decoded_content'][index[0]],'stemming')#movie

'crimin justic system peopl repres two separ yet equal import group polic investig crime district attorney prosecut offend stori api opensubtitl org deprec pleas implement rest api opensubtitl com yo yo hey yo run hey earli bird runner found lay approxim time death mlis put somewher wit none far yeah nobodi see noth huh eh blunt forc trauma guess id derek stanton still got cash indic bodi move way blood pool spatter pavement think happen right boy quit rap sheet got fishkil nine month ago sell marijuana serv four year four year sell weed crazi time chang surveil footag park near crime scene show guy wail stanton captur whole thing damn better video video got mani camera area even make face even skin color clear assail stanton size look guy feet tall els know well think perp might wear ring kind stone run vic bodi dna yeah took sampl backlog week minimum right well keep dig know given guy histori could relat drug mayb even time fishkil hey found someth interest vic phone past week stant

In [69]:
text_preprocessing(df['decoded_content'][index[2]],'stemming')#anime

'look elazul sea yeah pretti first time see sea seen distanc close shall take break big sis sens big sis glimmer yeah sens faint shiloh let go beach okay let fad bord fsp diamond pretti pearl crab look nlast time shiloh yeah went polpota big sis well esmeralda yeah think around somewher definit closer think sand probabl narrow search let walk bit okay well way think dialogu main top shiloh okay dialogu main top shiloh let go way shiloh kun voic shiloh kun valeri know yeah gon na go say hi long time valeri come swim friend travel search someth egg huh yes congratul could egg fruit love david dialogu main esmeralda seraphina david yes david belov valeri betray captain return ship belov david sorri freedom na lowli pirat like right go go defeat everi last monster nso never attack sea david make even difficult stop valeri come back mayb ten year twenti year hell might hundr year come back captain nof hundr penguin pirat long belov david warm egg may happi hope david get home yes noth fell 

### Observation
- It's working for both anime and non-anime transcripts

## Step-2.7 Check whether only these two patterns exist in the transcripts in the entire dataframe

In [70]:
def check_text(corpus):
    if re.search(dialogue_pattern,corpus)!=None or re.search(timestamp_pattern,corpus)!=None:
        return False
    return True

#df[df['decoded_content'].apply(check_text)]
# The below dataframe contains the transcripts which are not satisfied by current patterns
outlier_df=pd.read_csv('../data/temp.csv',index_col=0)
outlier_df

num                                               name  \
index                                                               
1128   9186306       skam.france.s10.e01.le.retour.(2022).eng.1cd   
1515   9187929  everythings.trash.s01.e04.public.image.is.tras...   
2221   9190728       hyper.projection.play.haikyuu.(2015).eng.1cd   
2291   9191497  band.of.brothers.s01.e07.the.breaking.point.(2...   
3690   9197707                           el.cortez.(2006).eng.1cd   
...        ...                                                ...   
79580  9509850  dimension.20.s07.e16.a.starstruck.odyssey.jail...   
79581  9509851  dimension.20.s07.e17.a.starstruck.odyssey.the....   
79582  9509852  dimension.20.s07.e18.a.starstruck.odyssey.the....   
80705  9514049                  shadow.of.the.wolf.(1992).eng.1cd   
81450  9517373  only.murders.in.the.building.s01.e07.the.boy.f...   

                                         decoded_content  
index                                                     
1128   [Script Info]\r\n; Script generated by Aegi...  
1515   WEBVTT\n\n00:02.737 --> 00:04.036\nOkay, but s...  
2221   [Script Info]\r\n; Script generated by Aegi...  
2291   {189}{254}I've seen death, I've seen...\r\n{27...  
3690   00:00:07,659 --> 00:00:08,688\r\n(piercing ...  
...                                                  ...  
79580  WEBVTT\r\n\r\n1\r\n00:00:02.531 --> 00:0...  
79581  WEBVTT\r\n\r\n1\r\n00:00:02.469 --> 00:0...  
79582  WEBVTT\r\n\r\n1\r\n00:00:00.000 --> 00:0...  
80705  {2555}{2668}GREAT WHITE NORTH|1935\r\n{7542}{7...  
81450  \n00:00:01,960 --> 00:00:04,379\nMABEL:\n<i>Zo...  

[226 rows x 3 columns]

### Observation
- As number of outliers is only 226 rows, it will be nor worth to modify pattern just to satisfy following transcripts. So ignore these outliers

## Step-3 Creating a Pre-processing Pipeline

## Step-3.1 Import Libraries

In [99]:
import nltk
from nltk.tokenize import word_tokenize
from nltk.stem import SnowballStemmer
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
import re
import inflect
from tqdm import tqdm
import pandas as pd

# Download NLTK resources if not already downloaded
nltk.download('punkt')
nltk.download('wordnet')
nltk.download('stopwords')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\dsai9\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\dsai9\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\dsai9\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

## Step-3.2 Data Ingestion

In [102]:
df=pd.read_csv('../data/eng_subtitles.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 82498 entries, 0 to 82497
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   num              82498 non-null  int64 
 1   name             82498 non-null  object
 2   decoded_content  82498 non-null  object
dtypes: int64(1), object(2)
memory usage: 1.9+ MB


In [103]:
#df=df.drop('num',axis=1)
#df['index']=df.index
#df=df.set_index('index')
df

name  \
index                                                      
0                             the.message.(1976).eng.1cd   
1      here.comes.the.grump.s01.e09.joltin.jack.in.bo...   
2        yumis.cells.s02.e13.episode.2.13.(2022).eng.1cd   
3        yumis.cells.s02.e14.episode.2.14.(2022).eng.1cd   
4                                  broker.(2022).eng.1cd   
...                                                  ...   
82493                   the.prophets.game.(2000).eng.1cd   
82494                         west.beirut.(1998).eng.1cd   
82495         frankenstein.the.true.story.(1973).eng.1cd   
82496         frankenstein.the.true.story.(1973).eng.1cd   
82497              zombie.island.massacre.(1984).eng.1cd   

                                         decoded_content  
index                                                     
0      1\r\n00:00:06,000 --> 00:00:12,074\r\nWatch an...  
1      1\r\n00:00:29,359 --> 00:00:32,048\r\nAh! Ther...  
2      1\r\n00:00:53,200 --> 00:00:56,030\r\n<i>Yumi'...  
3      1\r\n00:00:06,000 --> 00:00:12,074\r\nWatch an...  
4      1\r\n00:00:06,000 --> 00:00:12,074\r\nWatch...  
...                                                  ...  
82493  1\r\n00:01:16,284 --> 00:01:19,537\r\nGod,\...  
82494  1\r\n00:00:06,000 --> 00:00:12,074\r\napi.Open...  
82495  1\r\n00:00:01,001 --> 00:00:04,630\r\n(Dramati...  
82496  1\r\n00:00:06,000 --> 00:00:12,074\r\nAdvertis...  
82497  1\r\n00:00:01,919 --> 00:00:03,253\r\n(Sharp w...  

[82498 rows x 2 columns]

## Step-3.3 Pre processing the content

In [100]:
timestamp_pattern=r'\d+\r?\n\d{2}:\d{2}:\d{2},\d{3} --> \d{2}:\d{2}:\d{2},\d{3}'
dialogue_pattern=r"Dialogue: (Marked=)?\d+,\d{1,2}:\d{2}:\d{2}\.\d{2},\d{1,2}:\d{2}:\d{2}\.\d{2},\*?(\w+)?,(\w+)?,\d{1,4},\d{1,4},\d{1,4},,"
header_pattern1=r'Watch any video online with Open-SUBTITLES\r\nFree Browser extension: osdb.link/ext\r\n\r\n\r\n'#header
header_pattern2=r'\r\nAdvertise your product or brand here\r\ncontact www.OpenSubtitles.org today'
footer_pattern=r'Please rate this subtitle at www.osdb.link/[0-9a-z]+\r\nHelp other users to choose the best subtitles' #footer
header_pattern = r'\[Script Info\].*?\[Events\]\r\n.*?\r\n(Subtitle:)?'
link_pattern=r'(https?:\/\/)?(www\.)?(\w+)(\.\w+)'
linebreak_pattern=r'(\r?\n)+'
punctuation_pattern=r'[^a-zA-Z]'

In [101]:

def text_preprocessing(corpus,flag):


    if re.search(timestamp_pattern,corpus): #movie or tv show
        #remove timestamps
        corpus=re.sub(timestamp_pattern,'',corpus)
        #remove header and footer
        corpus=re.sub(header_pattern1,'',corpus)
        corpus=re.sub(footer_pattern,'',corpus)

    else: #anime
        #removing header
        corpus=re.sub(header_pattern,'',corpus,flags=re.DOTALL)
        #removing dialogue
        corpus=re.sub(dialogue_pattern,'',corpus)
        
    #remove linebreaks
    corpus=re.sub(linebreak_pattern,' ',corpus)
    
    #remove links
    corpus=re.sub(link_pattern,'',corpus)
    
    #remove punctuation
    corpus=re.sub(punctuation_pattern,' ',corpus)
    
    
    
    #convert to lower case
    corpus=corpus.lower()

    if flag == "stemming":
        stemmer = SnowballStemmer(language='english')
        preprocess_func = stemmer.stem
    else:
        lemmatizer = WordNetLemmatizer()
        preprocess_func = lemmatizer.lemmatize
    
    # Tokenize text
    words = word_tokenize(corpus.lower())
    
    # Precompute stopwords
    stop_words = set(nltk.corpus.stopwords.words('english'))
    
    # Process tokens
    processed_words = [preprocess_func(word) for word in words if word not in stop_words]
    
    # Join processed words
    return ' '.join(processed_words)

In [8]:
#df['pre-processed_content']=df['content'].apply(lambda x:text_preprocessing(x,'stemming'))
df

In [ ]:
#Save the file